### PR TITLE
Network and process states

### PR DESCRIPTION
**Pull request by trustmaster**
Implements #370 by making sure the following events are emitted at correct points of time:
- `start` - when the Network or Component starts accepting IPs
- `activate` - when a Component starts processing an incoming IP
- `deactivate` - when a Component finishes current IP processing cycle
- `end` - when a Component has stopped processing all current IPs and has safely shut down, or when a Network has safely shut down all of its components

This still to do / think over:
- [ ] Now Network emits `start` / `end` and also `Network.start(callback)` and `Network.stop(callback)` call the `callback` at the correct time, but the runtime is probably not utilizing these changes.
- [x] Improve API to notify the runtime of these events manually (e.g. in Generator components)
- [ ] Make sure queued input doesn't create any side effects and untracked issues

### CODE REVIEW

**Comment by trustmaster** [comment]
@bergie updated

File: spec/Component.coffee [hunk 1]
```diff
@@ -896,8 +896,8 @@ describe 'Component', ->
 
         c.inPorts.in.attach sin1
         sin1.post new noflo.IP 'data', 'some-data'
-      catch e
-        done()
+      chai.expect(f).to.thow
```
Comment (bergie) [inline spec/Component.coffee]: Typo here? Should be `throw`

**Comment by bergie** [comment]
@trustmaster apart from the typo above, do you feel this is good to go?

**Comment by trustmaster** [comment]
@bergie fixed and good to go 👍

**Pull request merged**